File: tests/parser_tests.rs
// Integration tests for the SQL query optimizer parser layer.
// These live in tests/ so they compile as a separate crate that imports
// sql_query_optimizer as a library — exactly how an external user would.

use sql_query_optimizer::parser::ast::*;
use sql_query_optimizer::parser::lexer::Lexer;
use sql_query_optimizer::parser::parser::Parser;

// ── helpers ──────────────────────────────────────────────────────────────────

fn parse(sql: &str) -> Statement {
    let tokens = Lexer::new(sql).tokenize().expect("lex error");
    Parser::new(tokens).parse().expect("parse error")
}

fn parse_err(sql: &str) -> String {
    let tokens = Lexer::new(sql).tokenize().expect("lex error");
    Parser::new(tokens).parse().expect_err("expected a parse error but got Ok")
}

fn select(sql: &str) -> SelectStatement {
    match parse(sql) {
        Statement::Select(s) => s,
    }
}

fn where_expr(sql: &str) -> Expression {
    select(sql).where_clause.expect("no WHERE clause")
}

// ── Lexer tests ───────────────────────────────────────────────────────────────

#[test]
fn lex_keywords_are_case_insensitive() {
    // "select" and "SELECT" should produce the same token stream
    let lower = Lexer::new("select * from t").tokenize().unwrap();
    let upper = Lexer::new("SELECT * FROM t").tokenize().unwrap();
    assert_eq!(lower, upper);
}

#[test]
fn lex_string_literal() {
    let tokens = Lexer::new("'hello world'").tokenize().unwrap();
    use sql_query_optimizer::parser::lexer::Token;
    assert!(matches!(&tokens[0], Token::StringLit(s) if s == "hello world"));
}

#[test]
fn lex_float_vs_integer() {
    use sql_query_optimizer::parser::lexer::Token;
    let tokens = Lexer::new("42 3.14").tokenize().unwrap();
    assert!(matches!(tokens[0], Token::Integer(42)));
    assert!(matches!(tokens[1], Token::Float(f) if (f - 3.14).abs() < 1e-9));
}

#[test]
fn lex_all_comparison_operators() {
    use sql_query_optimizer::parser::lexer::Token;
    let tokens = Lexer::new("= <> != < > <= >=").tokenize().unwrap();
    // Token::Eof at the end; 7 operators expected before it
    assert_eq!(tokens.len(), 8);
    assert!(matches!(tokens[0], Token::Eq));
    assert!(matches!(tokens[1], Token::Neq));
    assert!(matches!(tokens[2], Token::Neq)); // != is an alias for <>
    assert!(matches!(tokens[3], Token::Lt));
    assert!(matches!(tokens[4], Token::Gt));
    assert!(matches!(tokens[5], Token::Lte));
    assert!(matches!(tokens[6], Token::Gte));
}

// ── Parser — structural tests ─────────────────────────────────────────────────

#[test]
fn parse_select_star() {
    let s = select("SELECT * FROM orders");
    assert!(matches!(s.columns[0], SelectItem::Wildcard));
    assert_eq!(s.from.table_name, "orders");
    assert!(s.joins.is_empty());
    assert!(s.where_clause.is_none());
}

#[test]
fn parse_table_alias() {
    let s = select("SELECT * FROM customers c");
    assert_eq!(s.from.alias.as_deref(), Some("c"));
}

#[test]
fn parse_column_alias_with_as() {
    let s = select("SELECT price AS p FROM products");
    if let SelectItem::Expression(_, alias) = &s.columns[0] {
        assert_eq!(alias.as_deref(), Some("p"));
    } else {
        panic!("expected SelectItem::Expression");
    }
}

#[test]
fn parse_multiple_columns() {
    let s = select("SELECT id, name, email FROM users");
    assert_eq!(s.columns.len(), 3);
}

#[test]
fn parse_qualified_wildcard() {
    // t.* should become SelectItem::QualifiedWildcard("t")
    let s = select("SELECT t.* FROM t");
    assert!(
        matches!(&s.columns[0], SelectItem::QualifiedWildcard(tbl) if tbl == "t"),
        "expected QualifiedWildcard"
    );
}

#[test]
fn parse_left_join() {
    let s = select("SELECT * FROM a LEFT JOIN b ON a.id = b.a_id");
    assert_eq!(s.joins[0].join_type, JoinType::Left);
    assert_eq!(s.joins[0].table.table_name, "b");
}

#[test]
fn parse_right_join() {
    let s = select("SELECT * FROM a RIGHT JOIN b ON a.id = b.a_id");
    assert_eq!(s.joins[0].join_type, JoinType::Right);
}

#[test]
fn parse_cross_join() {
    let s = select("SELECT * FROM a CROSS JOIN b ON a.id = b.id");
    assert_eq!(s.joins[0].join_type, JoinType::Cross);
}

#[test]
fn parse_multiple_joins() {
    let sql = "SELECT * FROM a \
               INNER JOIN b ON a.id = b.a_id \
               LEFT JOIN c ON b.id = c.b_id";
    let s = select(sql);
    assert_eq!(s.joins.len(), 2);
    assert_eq!(s.joins[0].join_type, JoinType::Inner);
    assert_eq!(s.joins[1].join_type, JoinType::Left);
}

#[test]
fn parse_where_simple_equality() {
    let expr = where_expr("SELECT * FROM t WHERE status = 'active'");
    assert!(matches!(expr, Expression::BinaryOp { op: BinaryOperator::Eq, .. }));
}

#[test]
fn parse_group_by_multiple_columns() {
    let s = select("SELECT dept, role, COUNT(*) FROM emp GROUP BY dept, role");
    assert_eq!(s.group_by.len(), 2);
}

#[test]
fn parse_order_by_asc_default() {
    // Without ASC/DESC, should default to ascending
    let s = select("SELECT id FROM t ORDER BY id");
    assert!(s.order_by[0].ascending);
}

#[test]
fn parse_order_by_desc() {
    let s = select("SELECT id FROM t ORDER BY id DESC");
    assert!(!s.order_by[0].ascending);
}

#[test]
fn parse_limit() {
    let s = select("SELECT * FROM t LIMIT 25");
    assert_eq!(s.limit, Some(25));
}

#[test]
fn parse_no_limit_is_none() {
    let s = select("SELECT * FROM t");
    assert_eq!(s.limit, None);
}

#[test]
fn parse_function_avg() {
    let s = select("SELECT AVG(price) FROM products");
    if let SelectItem::Expression(Expression::Function { name, args }, _) = &s.columns[0] {
        assert_eq!(name, "AVG");
        assert_eq!(args.len(), 1);
    } else {
        panic!("expected Function expression");
    }
}

#[test]
fn parse_count_star() {
    let s = select("SELECT COUNT(*) FROM t");
    if let SelectItem::Expression(Expression::Function { name, args }, _) = &s.columns[0] {
        assert_eq!(name, "COUNT");
        assert!(matches!(args[0], Expression::Wildcard));
    } else {
        panic!("expected COUNT(*) function");
    }
}

// ── Parser — literal types ─────────────────────────────────────────────────────

#[test]
fn parse_integer_literal_in_where() {
    let expr = where_expr("SELECT 1 FROM t WHERE id = 99");
    if let Expression::BinaryOp { right, .. } = expr {
        assert!(matches!(*right, Expression::Literal(Literal::Integer(99))));
    }
}

#[test]
fn parse_string_literal_in_where() {
    let expr = where_expr("SELECT 1 FROM t WHERE name = 'alice'");
    if let Expression::BinaryOp { right, .. } = expr {
        assert!(matches!(*right, Expression::Literal(Literal::Str(ref s)) if s == "alice"));
    }
}

#[test]
fn parse_boolean_true() {
    let expr = where_expr("SELECT 1 FROM t WHERE active = TRUE");
    if let Expression::BinaryOp { right, .. } = expr {
        assert!(matches!(*right, Expression::Literal(Literal::Boolean(true))));
    }
}

#[test]
fn parse_null_literal() {
    let expr = where_expr("SELECT 1 FROM t WHERE deleted = NULL");
    if let Expression::BinaryOp { right, .. } = expr {
        assert!(matches!(*right, Expression::Literal(Literal::Null)));
    }
}

// ── Parser — expression precedence ────────────────────────────────────────────

#[test]
fn and_binds_tighter_than_or() {
    // a=1 OR b=2 AND c=3  =>  OR(a=1, AND(b=2, c=3))
    let expr = where_expr("SELECT 1 FROM t WHERE a = 1 OR b = 2 AND c = 3");
    assert!(
        matches!(expr, Expression::BinaryOp { op: BinaryOperator::Or, .. }),
        "top-level must be OR"
    );
}

#[test]
fn mul_binds_tighter_than_add() {
    // a + b*c = 0  =>  Eq(Add(a, Mul(b,c)), 0)
    let expr = where_expr("SELECT 1 FROM t WHERE a + b * c = 0");
    if let Expression::BinaryOp { op: BinaryOperator::Eq, left, .. } = expr {
        assert!(matches!(*left, Expression::BinaryOp { op: BinaryOperator::Add, .. }));
    }
}

#[test]
fn parens_override_precedence() {
    // (a+b)*c = 0  =>  Eq(Mul(Add(a,b),c), 0)
    let expr = where_expr("SELECT 1 FROM t WHERE (a + b) * c = 0");
    if let Expression::BinaryOp { op: BinaryOperator::Eq, left, .. } = expr {
        assert!(matches!(*left, Expression::BinaryOp { op: BinaryOperator::Mul, .. }));
    }
}

#[test]
fn not_binds_tighter_than_and() {
    // NOT a=1 AND b=2  =>  AND(NOT(a=1), b=2)
    let expr = where_expr("SELECT 1 FROM t WHERE NOT a = 1 AND b = 2");
    if let Expression::BinaryOp { op: BinaryOperator::And, left, .. } = expr {
        assert!(matches!(*left, Expression::UnaryOp { op: UnaryOperator::Not, .. }));
    }
}

#[test]
fn unary_minus_precedence() {
    // -a + b = 0  =>  Eq(Add(Neg(a), b), 0)
    let expr = where_expr("SELECT 1 FROM t WHERE -a + b = 0");
    if let Expression::BinaryOp { op: BinaryOperator::Eq, left, .. } = expr {
        if let Expression::BinaryOp { op: BinaryOperator::Add, left: neg, .. } = *left {
            assert!(matches!(*neg, Expression::UnaryOp { op: UnaryOperator::Neg, .. }));
        }
    }
}

#[test]
fn chained_and() {
    // a=1 AND b=2 AND c=3 is left-associative: AND(AND(a=1,b=2), c=3)
    let expr = where_expr("SELECT 1 FROM t WHERE a = 1 AND b = 2 AND c = 3");
    if let Expression::BinaryOp { op: BinaryOperator::And, left, .. } = expr {
        assert!(matches!(*left, Expression::BinaryOp { op: BinaryOperator::And, .. }));
    } else {
        panic!("expected AND at top");
    }
}

// ── Print / inspect tests  (run with: cargo test print_ -- --nocapture) ────────

#[test]
fn print_ast_group_by_sum() {
    let sql = "SELECT c.name, SUM(o.total) \
               FROM customers c \
               JOIN orders o ON c.id = o.customer_id \
               GROUP BY c.name";

    let stmt = parse(sql);
    println!("SQL: {}", sql);
    println!("AST: {:#?}", stmt);
}

// ── Parser — error cases ───────────────────────────────────────────────────────

#[test]
fn error_on_missing_from() {
    let err = parse_err("SELECT *");
    assert!(!err.is_empty(), "should return an error message");
}

#[test]
fn error_on_empty_input() {
    let err = parse_err("");
    assert!(!err.is_empty());
}

#[test]
fn error_on_non_select() {
    // Only SELECT is supported so far; INSERT should fail
    let err = parse_err("INSERT INTO t VALUES (1)");
    assert!(!err.is_empty());
}

File: tests/parser_tests.cpp
// Tests for the SQL query optimizer parser layer (lexer + recursive-descent
// parser). Ported 1:1 from the original Rust `tests/parser_tests.rs`.

#include <cmath>
#include <cstdint>
#include <stdexcept>
#include <string>

#include "../parser/ast.hpp"
#include "../parser/lexer.hpp"
#include "../parser/parser.hpp"
#include "test_framework.hpp"

using namespace sql::parser;

namespace {

// ── helpers ──────────────────────────────────────────────────────────────────

Statement parse(const std::string& sql) {
    Lexer lexer(sql);
    std::vector<Token> tokens = lexer.tokenize();
    Parser parser(std::move(tokens));
    return parser.parse();
}

std::string parse_err(const std::string& sql) {
    Lexer lexer(sql);
    std::vector<Token> tokens;
    try {
        tokens = lexer.tokenize();
    } catch (const std::exception& e) {
        FAIL_TEST(std::string("lex error: ") + e.what());
    }
    Parser parser(std::move(tokens));
    try {
        parser.parse();
    } catch (const std::exception& e) {
        return e.what();
    }
    FAIL_TEST("expected a parse error but got Ok");
    return "";
}

SelectStatement select(const std::string& sql) {
    Statement stmt = parse(sql);
    return std::move(stmt.select);
}

Expression where_expr(const std::string& sql) {
    SelectStatement s = select(sql);
    ASSERT_TRUE_MSG(s.where_clause.has_value(), "no WHERE clause");
    return std::move(*s.where_clause);
}

} // namespace

// ── Lexer tests ────────────────────────────────────────────────────────────────

TEST(lex_keywords_are_case_insensitive) {
    // "select" and "SELECT" should produce the same token stream
    Lexer lower_lexer("select * from t");
    Lexer upper_lexer("SELECT * FROM t");
    auto lower = lower_lexer.tokenize();
    auto upper = upper_lexer.tokenize();
    ASSERT_TRUE(lower == upper);
}

TEST(lex_string_literal) {
    Lexer lexer("'hello world'");
    auto tokens = lexer.tokenize();
    ASSERT_TRUE(tokens[0].kind == TokenKind::StringLit);
    ASSERT_EQ(tokens[0].text(), std::string("hello world"));
}

TEST(lex_float_vs_integer) {
    Lexer lexer("42 3.14");
    auto tokens = lexer.tokenize();
    ASSERT_TRUE(tokens[0].kind == TokenKind::Integer);
    ASSERT_EQ(tokens[0].integer(), static_cast<int64_t>(42));
    ASSERT_TRUE(tokens[1].kind == TokenKind::Float);
    ASSERT_TRUE(std::abs(tokens[1].floating() - 3.14) < 1e-9);
}

TEST(lex_all_comparison_operators) {
    Lexer lexer("= <> != < > <= >=");
    auto tokens = lexer.tokenize();
    // Token::Eof at the end; 7 operators expected before it
    ASSERT_EQ(tokens.size(), static_cast<size_t>(8));
    ASSERT_TRUE(tokens[0].kind == TokenKind::Eq);
    ASSERT_TRUE(tokens[1].kind == TokenKind::Neq);
    ASSERT_TRUE(tokens[2].kind == TokenKind::Neq); // != is an alias for <>
    ASSERT_TRUE(tokens[3].kind == TokenKind::Lt);
    ASSERT_TRUE(tokens[4].kind == TokenKind::Gt);
    ASSERT_TRUE(tokens[5].kind == TokenKind::Lte);
    ASSERT_TRUE(tokens[6].kind == TokenKind::Gte);
}

// ── Parser -- structural tests ───────────────────────────────────────────────

TEST(parse_select_star) {
    auto s = select("SELECT * FROM orders");
    ASSERT_TRUE(s.columns[0].kind == SelectItem::Kind::Wildcard);
    ASSERT_EQ(s.from.table_name, std::string("orders"));
    ASSERT_TRUE(s.joins.empty());
    ASSERT_FALSE(s.where_clause.has_value());
}

TEST(parse_table_alias) {
    auto s = select("SELECT * FROM customers c");
    ASSERT_TRUE(s.from.alias.has_value());
    ASSERT_EQ(*s.from.alias, std::string("c"));
}

TEST(parse_column_alias_with_as) {
    auto s = select("SELECT price AS p FROM products");
    ASSERT_TRUE(s.columns[0].kind == SelectItem::Kind::Expression);
    ASSERT_TRUE(s.columns[0].alias.has_value());
    ASSERT_EQ(*s.columns[0].alias, std::string("p"));
}

TEST(parse_multiple_columns) {
    auto s = select("SELECT id, name, email FROM users");
    ASSERT_EQ(s.columns.size(), static_cast<size_t>(3));
}

TEST(parse_qualified_wildcard) {
    // t.* should become SelectItem::QualifiedWildcard("t")
    auto s = select("SELECT t.* FROM t");
    ASSERT_TRUE_MSG(s.columns[0].kind == SelectItem::Kind::QualifiedWildcard, "expected QualifiedWildcard");
    ASSERT_EQ(s.columns[0].qualified_table, std::string("t"));
}

TEST(parse_left_join) {
    auto s = select("SELECT * FROM a LEFT JOIN b ON a.id = b.a_id");
    ASSERT_TRUE(s.joins[0].join_type == JoinType::Left);
    ASSERT_EQ(s.joins[0].table.table_name, std::string("b"));
}

TEST(parse_right_join) {
    auto s = select("SELECT * FROM a RIGHT JOIN b ON a.id = b.a_id");
    ASSERT_TRUE(s.joins[0].join_type == JoinType::Right);
}

TEST(parse_cross_join) {
    auto s = select("SELECT * FROM a CROSS JOIN b ON a.id = b.id");
    ASSERT_TRUE(s.joins[0].join_type == JoinType::Cross);
}

TEST(parse_multiple_joins) {
    std::string sql = "SELECT * FROM a "
                       "INNER JOIN b ON a.id = b.a_id "
                       "LEFT JOIN c ON b.id = c.b_id";
    auto s = select(sql);
    ASSERT_EQ(s.joins.size(), static_cast<size_t>(2));
    ASSERT_TRUE(s.joins[0].join_type == JoinType::Inner);
    ASSERT_TRUE(s.joins[1].join_type == JoinType::Left);
}

TEST(parse_where_simple_equality) {
    auto expr = where_expr("SELECT * FROM t WHERE status = 'active'");
    ASSERT_TRUE(expr.kind == Expression::Kind::BinaryOp);
    ASSERT_TRUE(expr.binary_op == BinaryOperator::Eq);
}

TEST(parse_group_by_multiple_columns) {
    auto s = select("SELECT dept, role, COUNT(*) FROM emp GROUP BY dept, role");
    ASSERT_EQ(s.group_by.size(), static_cast<size_t>(2));
}

TEST(parse_order_by_asc_default) {
    // Without ASC/DESC, should default to ascending
    auto s = select("SELECT id FROM t ORDER BY id");
    ASSERT_TRUE(s.order_by[0].ascending);
}

TEST(parse_order_by_desc) {
    auto s = select("SELECT id FROM t ORDER BY id DESC");
    ASSERT_FALSE(s.order_by[0].ascending);
}

TEST(parse_limit) {
    auto s = select("SELECT * FROM t LIMIT 25");
    ASSERT_TRUE(s.limit.has_value());
    ASSERT_EQ(*s.limit, static_cast<size_t>(25));
}

TEST(parse_no_limit_is_none) {
    auto s = select("SELECT * FROM t");
    ASSERT_FALSE(s.limit.has_value());
}

TEST(parse_function_avg) {
    auto s = select("SELECT AVG(price) FROM products");
    ASSERT_TRUE(s.columns[0].kind == SelectItem::Kind::Expression);
    const Expression& e = *s.columns[0].expr;
    ASSERT_TRUE(e.kind == Expression::Kind::Function);
    ASSERT_EQ(e.func_name, std::string("AVG"));
    ASSERT_EQ(e.args.size(), static_cast<size_t>(1));
}

TEST(parse_count_star) {
    auto s = select("SELECT COUNT(*) FROM t");
    ASSERT_TRUE(s.columns[0].kind == SelectItem::Kind::Expression);
    const Expression& e = *s.columns[0].expr;
    ASSERT_TRUE(e.kind == Expression::Kind::Function);
    ASSERT_EQ(e.func_name, std::string("COUNT"));
    ASSERT_TRUE(e.args[0].kind == Expression::Kind::Wildcard);
}

// ── Parser -- literal types ───────────────────────────────────────────────────

TEST(parse_integer_literal_in_where) {
    auto expr = where_expr("SELECT 1 FROM t WHERE id = 99");
    ASSERT_TRUE(expr.kind == Expression::Kind::BinaryOp);
    ASSERT_TRUE(expr.right->kind == Expression::Kind::Literal);
    ASSERT_TRUE(expr.right->literal.kind == Literal::Kind::Integer);
    ASSERT_EQ(expr.right->literal.int_val, static_cast<int64_t>(99));
}

TEST(parse_string_literal_in_where) {
    auto expr = where_expr("SELECT 1 FROM t WHERE name = 'alice'");
    ASSERT_TRUE(expr.right->kind == Expression::Kind::Literal);
    ASSERT_TRUE(expr.right->literal.kind == Literal::Kind::Str);
    ASSERT_EQ(expr.right->literal.str_val, std::string("alice"));
}

TEST(parse_boolean_true) {
    auto expr = where_expr("SELECT 1 FROM t WHERE active = TRUE");
    ASSERT_TRUE(expr.right->kind == Expression::Kind::Literal);
    ASSERT_TRUE(expr.right->literal.kind == Literal::Kind::Boolean);
    ASSERT_TRUE(expr.right->literal.bool_val);
}

TEST(parse_null_literal) {
    auto expr = where_expr("SELECT 1 FROM t WHERE deleted = NULL");
    ASSERT_TRUE(expr.right->kind == Expression::Kind::Literal);
    ASSERT_TRUE(expr.right->literal.kind == Literal::Kind::Null);
}

// ── Parser -- expression precedence ──────────────────────────────────────────

TEST(and_binds_tighter_than_or) {
    // a=1 OR b=2 AND c=3  =>  OR(a=1, AND(b=2, c=3))
    auto expr = where_expr("SELECT 1 FROM t WHERE a = 1 OR b = 2 AND c = 3");
    ASSERT_TRUE_MSG(expr.kind == Expression::Kind::BinaryOp && expr.binary_op == BinaryOperator::Or,
                     "top-level must be OR");
}

TEST(mul_binds_tighter_than_add) {
    // a + b*c = 0  =>  Eq(Add(a, Mul(b,c)), 0)
    auto expr = where_expr("SELECT 1 FROM t WHERE a + b * c = 0");
    ASSERT_TRUE(expr.kind == Expression::Kind::BinaryOp && expr.binary_op == BinaryOperator::Eq);
    ASSERT_TRUE(expr.left->kind == Expression::Kind::BinaryOp && expr.left->binary_op == BinaryOperator::Add);
}

TEST(parens_override_precedence) {
    // (a+b)*c = 0  =>  Eq(Mul(Add(a,b),c), 0)
    auto expr = where_expr("SELECT 1 FROM t WHERE (a + b) * c = 0");
    ASSERT_TRUE(expr.kind == Expression::Kind::BinaryOp && expr.binary_op == BinaryOperator::Eq);
    ASSERT_TRUE(expr.left->kind == Expression::Kind::BinaryOp && expr.left->binary_op == BinaryOperator::Mul);
}

TEST(not_binds_tighter_than_and) {
    // NOT a=1 AND b=2  =>  AND(NOT(a=1), b=2)
    auto expr = where_expr("SELECT 1 FROM t WHERE NOT a = 1 AND b = 2");
    ASSERT_TRUE(expr.kind == Expression::Kind::BinaryOp && expr.binary_op == BinaryOperator::And);
    ASSERT_TRUE(expr.left->kind == Expression::Kind::UnaryOp && expr.left->unary_op == UnaryOperator::Not);
}

TEST(unary_minus_precedence) {
    // -a + b = 0  =>  Eq(Add(Neg(a), b), 0)
    auto expr = where_expr("SELECT 1 FROM t WHERE -a + b = 0");
    ASSERT_TRUE(expr.kind == Expression::Kind::BinaryOp && expr.binary_op == BinaryOperator::Eq);
    const Expression& add = *expr.left;
    ASSERT_TRUE(add.kind == Expression::Kind::BinaryOp && add.binary_op == BinaryOperator::Add);
    ASSERT_TRUE(add.left->kind == Expression::Kind::UnaryOp && add.left->unary_op == UnaryOperator::Neg);
}

TEST(chained_and) {
    // a=1 AND b=2 AND c=3 is left-associative: AND(AND(a=1,b=2), c=3)
    auto expr = where_expr("SELECT 1 FROM t WHERE a = 1 AND b = 2 AND c = 3");
    ASSERT_TRUE_MSG(expr.kind == Expression::Kind::BinaryOp && expr.binary_op == BinaryOperator::And,
                     "expected AND at top");
    ASSERT_TRUE(expr.left->kind == Expression::Kind::BinaryOp && expr.left->binary_op == BinaryOperator::And);
}

// ── Print / inspect (smoke test) ──────────────────────────────────────────────

TEST(print_ast_group_by_sum) {
    std::string sql = "SELECT c.name, SUM(o.total) "
                       "FROM customers c "
                       "JOIN orders o ON c.id = o.customer_id "
                       "GROUP BY c.name";
    // Just confirm this parses without throwing.
    Statement stmt = parse(sql);
    (void)stmt;
}

// ── Parser -- error cases ─────────────────────────────────────────────────────

TEST(error_on_missing_from) {
    auto err = parse_err("SELECT *");
    ASSERT_TRUE_MSG(!err.empty(), "should return an error message");
}

TEST(error_on_empty_input) {
    auto err = parse_err("");
    ASSERT_FALSE(err.empty());
}

TEST(error_on_non_select) {
    // Only SELECT is supported so far; INSERT should fail
    auto err = parse_err("INSERT INTO t VALUES (1)");
    ASSERT_FALSE(err.empty());
}
